Fix local likes/boosts not sending notifications to post creators

ivanminutillo opened the pull request:
## Summary
This PR fixes an issue where local users liking or boosting posts would not send notifications to the post creator, while remote (federated) users would correctly trigger notifications.
addresses https://github.com/bonfire-networks/bonfire-app/issues/1222


## Root Cause
The issue occurred because:
1. Posts were being loaded as `Needle.Pointer` objects without proper resolution
2. The `Created` mixin association was `nil` on Post objects
3. The system couldn't find the creator, resulting in empty `to_feeds` and no notifications

## Changes Made

### Core Fix in `objects.ex`
- Enhanced `preload_creator/1` to properly resolve `Needle.Pointer` objects via `:pointed` or `:post` associations
- Added fallback logic to load creator from the activity's subject when the `Created` mixin is missing
- Enhanced `object_creator/1` to check activity subject as a fallback for finding creators

### Bug Fixes
- Fixed variable shadowing bug in `likes.ex`: `opts` → `new_opts` in `do_like/3`
- Added nil check in `feeds.ex` `maybe_creator_notification/3` to prevent boolean errors
- Added defensive fallback logic in both likes and boosts for creator resolution

### Minor Changes
- Added necessary module aliases
- Added debug logging for troubleshooting

## Expected Behavior
- Local users liking/boosting posts should trigger notifications to post creators
- Users should not receive notifications for their own actions
- Remote users should continue to work as before (no regression)

## Related Issues
This issue was discovered when local likes/boosts weren't appearing in the notifications feed, while federated interactions worked correctly.

ivanminutillo committed:
Fix local likes/boosts not sending notifications to post creators

- Enhanced creator resolution in Objects.preload_creator/1 to handle Needle.Pointers and missing Created associations
- Added fallback to load creator from activity subject when Created mixin is nil
- Fixed variable shadowing bug in Likes.do_like/3 (opts -> new_opts)
- Added nil check in Feeds.maybe_creator_notification/3 to prevent crashes
- Added defensive fallbacks in likes and boosts for creator resolution

Fixes issue where local users liking/boosting posts wouldn't notify the post creator, while remote users would.

mayel commented on lib/likes.ex:
```diff
@@ -195,18 +192,41 @@ defmodule Bonfire.Social.Likes do
     liked_object_creator = Objects.object_creator(liked)
 
-    opts =
+    # Fallback: if no creator found, try to get it from activity or other associations
```
it seems this is repeating logic that's also in Boosts? could this move to `Objects.object_creator/1` since both are using it?

mayel replied:
or maybe creating an Activities.activity_creator? and call one or the other depending on if we have an activity or object, rather than both

mayel commented on lib/objects.ex:
```diff
@@ -210,11 +210,70 @@ defmodule Bonfire.Social.Objects do
+    object = case object do
+      %Needle.Pointer{} = pointer ->
+        debug("Resolving Needle.Pointer")
+        # Need to properly resolve the pointer to the actual object
```
why?

mayel commented on lib/objects.ex:
```diff
@@ -210,11 +210,70 @@ defmodule Bonfire.Social.Objects do
+      %Needle.Pointer{} = pointer ->
+        debug("Resolving Needle.Pointer")
+        # Need to properly resolve the pointer to the actual object
+        resolved = repo().maybe_preload(pointer, :pointed)
```
we have reusable logic for this stuff in Bonfire.Common.Needles and should avoid polluting the rest of the codebase with too needle/pointer implementation specific

mayel commented on lib/objects.ex:
```diff
@@ -210,11 +210,70 @@ defmodule Bonfire.Social.Objects do
+        # Check if this is a create activity
+        if verb_id == Bonfire.Data.AccessControl.Verbs.get_id!(:create) do
+          debug("Loading creator from create activity subject")
+          repo().maybe_preload(with_activity, activity: [subject: [:character]])
```
so many preloads in this function! are they all needed? can't they be done all at once?

mayel commented on lib/objects.ex:
```diff
@@ -226,7 +285,29 @@ defmodule Bonfire.Social.Objects do
+    # Try standard paths first
+    creator = e(object, :created, :creator, nil) || e(object, :creator, nil)
+    
+    # If no creator found and we have an activity, check if it's a create activity
```
is this duplicating code also in Boosts and Likes?

mayel commented on lib/likes.ex:
```diff
@@ -195,18 +192,41 @@ defmodule Bonfire.Social.Likes do
+
+    debug("do_like: found creator #{inspect(id(liked_object_creator))}")
+
+    new_opts =
```
no need to call it something else, put actually would be cleaner to have opts = opts |> Keyword.merge(...

mayel commented on lib/feeds.ex:
```diff
@@ -734,11 +734,18 @@ defmodule Bonfire.Social.Feeds do
+      debug("Creator notification: no creator found, returning empty list")
       []
+    else
+      if id(subject) != id(object_creator) and
```
not sure what this is doing?

mayel commented on lib/boosts.ex:
```diff
@@ -178,9 +179,33 @@ defmodule Bonfire.Social.Boosts do
+    debug("do_boost: booster=#{id(booster)}, creator=#{inspect(id(boosted_creator))}")
+
+    # Use "public_remote" boundary for federated boosts
+    boundary = if e(opts, :local, true), do: "public", else: "public_remote"
```
good catch!

mayel reviewed (commented):
added some initial comments, still need to dig in though

ivanminutillo committed:
Refactor creator resolution based on PR feedback

Addresses reviewer feedback from @mayel to improve code quality and reduce duplication:

1. **Consolidate duplicated creator resolution logic**
   - Moved fallback logic from Likes.do_like/3 and Boosts.do_boost/3 into Objects.object_creator/1
   - Eliminates code duplication between like and boost modules
   - Creates single source of truth for creator resolution edge cases

2. **Use existing Needles infrastructure**
   - Replaced custom pointer resolution with Bonfire.Common.Needles.follow\!/1
   - Avoids "polluting the codebase with needle/pointer implementation specific" code
   - Leverages well-tested framework functionality

3. **Optimize preloading efficiency**
   - Combined multiple preload calls into single repo().maybe_preload/2 operation
   - Addresses concern about "so many preloads in this function"
   - Reduces potential N+1 query issues

4. **Improve variable naming**
   - Changed `new_opts` to `opts |> Keyword.merge(...)` pattern
   - Makes code more idiomatic and concise
   - Follows Elixir community conventions

5. **Remove excessive debug logging**
   - Commented out verbose debug statements for production readiness
   - Keeps essential logging while reducing noise

The consolidated object_creator/1 function now handles all edge cases:
- Activities (distinguishing create vs other verbs)
- Objects with missing Created associations
- Various fallback paths (activity.subject, post_content.created.creator)

This maintains the original PR's functionality (fixing local notification issues)
while addressing all architectural concerns raised in the review.

ivanminutillo committed:
right icon

mayel commented on lib/objects.ex:
```diff
@@ -226,7 +242,39 @@ defmodule Bonfire.Social.Objects do
+              e(activity, :subject, nil)
+            else
+              # For other activities, try to get the object's creator
+              activity = repo().maybe_preload(activity, object: [created: [creator: :character]])
```
preload_creator is called before this function, so should not have to preload again

mayel commented on lib/objects.ex:
```diff
@@ -210,11 +210,27 @@ defmodule Bonfire.Social.Objects do
+    # Preload all creator-related associations at once
+    object
+    |> repo().maybe_preload(
+      created: [creator: [:character]],
```
why preload both the top-level object creator and the activity object creator? can't we do one or the other? maybe doing some pattern matching in function head to see if we have an activity or object?

mayel commented on lib/objects.ex:
```diff
@@ -210,11 +210,27 @@ defmodule Bonfire.Social.Objects do
+    # First resolve if it's a Pointer using Bonfire.Common.Needles
+    object =
+      case object do
+        %Needle.Pointer{} = pointer ->
```
just %Needle.Pointer{} is enought

mayel commented on lib/objects.ex:
```diff
@@ -210,11 +210,27 @@ defmodule Bonfire.Social.Objects do
+      case object do
+        %Needle.Pointer{} = pointer ->
+          # Use Bonfire.Common.Needles for proper pointer resolution
+          Common.Needles.follow!(pointer) || pointer
```
Common.Needles.follow!(object) || object

mayel commented on lib/objects.ex:
```diff
@@ -210,11 +210,27 @@ defmodule Bonfire.Social.Objects do
+          # Use Bonfire.Common.Needles for proper pointer resolution
+          Common.Needles.follow!(pointer) || pointer
+
+        other ->
```
_ ->
object

mayel commented:
merged and made some tweaks but need test to verify
